// jvelezpo-pager-cli/cli/cmds/get_cmds/auth-code.js
'use strict';

const JWT = require('jsonwebtoken');
const moment = require('moment');
const { v4: uuidv4 } = require('uuid');

exports.command = 'auth-code [email] [first-name] [last-name] [private-key]';

exports.desc = 'Auth code generator';

exports.builder = {
    'email': {
        required: true
    },
    'first-name': {
        required: true
    },
    'last-name': {
        required: true
    },
    'private-key': {
        required: true
    }
};

exports.handler = function (argv) {

    const { email, firstName, lastName, privateKey } = argv;
    const sub = uuidv4();

    const members = {
        'members': [
            {
                'memberId': String(sub),
                'externalId': String(sub),
                'subscriberId': String(sub),
                'firstName': firstName + '(' + sub + ')',
                lastName,
                'email': email + '+' + sub + '@pager.com',
                'dateOfBirth': '19900129',
                'gender': 'M'
            }
        ]
    };

    const payload = {
        sub: String(sub) // use new sub for every new user here, remember that all users should have a unique sub
    };

    const token = JWT.sign(payload, privateKey, { algorithm: 'ES256', keyid: 'abc123' });

    console.log('AuthCode:\n', token);
    console.log('\nData:\n');
    console.log(JSON.stringify(members));
};
