// jvelezpo-pager-cli/cli/cmds/kubectl/tunnel.js
'use strict';

const { execSync, spawnSync } = require('child_process');
const { validateNamespace, validateDb, validatePort } = require('../../../scripts/utils/validators');

exports.desc = 'create db tunnel';

exports.command = 'tunnel [namespace] [db] [port]';

exports.aliases = ['db-tunnel'];

exports.builder = {
    'namespace': {
        required: true,
        coerce: validateNamespace,
        alias: ['n', 'ns']
    },
    'db': {
        required: false,
        coerce: validateDb,
        alias: 'd'
    },
    'port': {
        required: false,
        coerce: validatePort,
        alias: 'p'
    }
};

exports.example = 'create db tunnel --namespace=pgr-qa --db=mongo --port=27017';

exports.handler = function handler(argv) {

    const { ns, db, port } = argv;

    // Get pod name
    console.log('Getting podname from K8s...');
    const podName = execSync(`kubectl get pods --namespace ${ns} -l "app=${db}" -o jsonpath="{.items[1].metadata.name}"`);

    // Port forward
    console.log(`Port forwarding localhost:${port} to ${db} on namespace ${ns}...`);
    const portFowardResult = spawnSync('kubectl', ['--namespace', ns, 'port-forward', podName, `${port}:27017`]);

    /*
    TODO:
     - spawn subshell to get env-* repo contents
     - decrypt config/secrets.yaml & extract pw (sops -d --extract '["mongodb-replicaset"]["auth"]["adminPassword"]' config/secrets.yaml)
     - create db tunnel string that includes the pw and port
     - actually run the tunnel command?
  */
};
